pkg/streamingpromql/optimize/plan/commonsubexpressionelimination/optimization_pass.go

```diff
@@ -0,0 +1,343 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package commonsubexpressionelimination
+
+import (
+	"context"
+	"slices"
+	"strings"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/prometheus/prometheus/promql/parser"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/planning"
+	"github.com/grafana/mimir/pkg/streamingpromql/planning/core"
+	"github.com/grafana/mimir/pkg/streamingpromql/types"
+)
+
+// This file implements common subexpression elimination.
+//
+// This allows us to skip evaluating the same expression multiple times.
+// For example, in the expression "sum(a) / (sum(a) + sum(b))", we can skip evaluating the "sum(a)" expression a second time
+// and instead reuse the result from the first evaluation.
+//
+// OptimizationPass is an optimization pass that identifies common subexpressions and injects Duplicate nodes
+// into the query plan where needed.
+//
+// When the query plan is materialized, a DuplicationBuffer is created to buffer the results of the common subexpression,
+// and a DuplicationConsumer is created for each consumer of the common subexpression.
+
+type OptimizationPass struct {
+	duplicationNodesIntroduced prometheus.Counter
+}
+
+func NewOptimizationPass(reg prometheus.Registerer) *OptimizationPass {
+	return &OptimizationPass{
+		duplicationNodesIntroduced: promauto.With(reg).NewCounter(prometheus.CounterOpts{
```

Comment:
what if other optimisations want to use the duplication node? Wouldn't the metric be better off held on the duplication node and then a label used to determine the source?

Reply:
For me, this metric is about evaluating the impact of this optimization pass, rather than the node. So if we end up using the node for other optimization passes, I'd expect that optimization pass to have its own metrics.